### PR TITLE
First cut of AIE Delegates

### PR DESCRIPTION
**Pull request by daveliddell**
An AIE Delegate is a C/C++ function implementing an operator (or fusion of operators) by dispatching to a pre-built AIE kernel.

Please see the README.md file included in this PR for details.

### CODE REVIEW

File: experimental/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Copyright 2022 The IREE Authors
```
Comment (nirvedhmeshram) [inline experimental/CMakeLists.txt:1]: nit: 2024

File: experimental/delegate/README.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+```
+
+### Run the model
+
+```
+# Circumvent xclbin security
+export XRT_HACK_UNSECURE_LOADING_XCLBIN=1
+
+iree-run-module --device=local-sync \
+  --executable_plugin=../../../iree-build/runtime/plugins/AMD-AIE-experimental/delegate/mlp_bf16_aie_delegate.so \
```
Comment (nirvedhmeshram) [inline experimental/delegate/README.md:34]: nit: say {$PATH_TO_IREE_BUILD}/runtime....

File: experimental/delegate/kernels/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+# Install each file to the destination directory and build a list of destination files
+set(AIE_DELEGATE_KERNEL_DEST_FILES)
+foreach(kernel ${AIE_DELEGATE_KERNELS})
+    foreach (extension ${AIE_DELEGATE_KERNEL_EXTENSIONS})
+        set(_src_file "${kernel}${extension}")
+        set(_dest_file "${CMAKE_CURRENT_BINARY_DIR}/${_src_file}")
+        set(_url "${AIE_DELEGATE_KERNEL_REPO}/${_src_file}")
+        add_custom_command(
+            OUTPUT ${_dest_file}
+            COMMAND curl ${_url} -o ${_dest_file}
```
Comment (nirvedhmeshram) [inline experimental/delegate/kernels/CMakeLists.txt:40]: I think we prefer wget over curl, but I am not an expert on why.

Comment (daveliddell) [reply experimental/delegate/kernels/CMakeLists.txt:40]: It doesn't look like either wget or curl is guaranteed to be installed on every Ubuntu 22.04.  In our case, wget is not installed on the CI machine, but curl is

File: experimental/CMakeLists.txt (hunk 1 above)
Comment (daveliddell) [reply experimental/CMakeLists.txt]: Fixed, with reservations.  I copied this file from elsewhere unchanged.  My understanding of copyright law is that the copyright date must not change unless the document does.  (But who's to say that I didn't write it from scratch!)  As I was told in a mandatory IP training class years ago, inconsistencies in the way that copyright claims are handled weakens your claim to the copyright in court.  I suspect that "IREE Authors," not being a legal entity, is also invalid, as there is no entity laying claim to the copyright.

File: experimental/delegate/README.md (hunk 2 above)
Comment (daveliddell) [reply experimental/delegate/README.md]: Fixed, but not exactly in the way requested.  I've kept the default path elsewhere in the instructions, as we have plans to turn our internal e2e IREE/AIE build doc into a public doc.  It sets up a specific directory structure for the sake of simplifying the build process for users.

**Comment by MaheshRavishankar** [comment]
Fly by comment. Please use better PR descriptions (believe it or not, people are following PRs being sent to this repro, and good PR hygiene matters a lot in OSS).

File: experimental/delegate/mlp.mlir [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+#cpu_target = #hal.device.target<"llvm-cpu", [
+  #x86_64_target
+]>
+
+#map = affine_map<(d0, d1) -> (d0, d1)>
+module @example attributes {hal.device.targets = [#cpu_target]} {
+
+  // CHECK-LABEL: EXEC @mlp_invocation
+  //       CHECK: [Plugin]: M = 2, N = 2, K = 2
+  //       CHECK: 2x2xf32=[-12 0][0 -12]
```
Comment (nirvedhmeshram) [inline experimental/delegate/mlp.mlir:30]: I am not following how we ended up with M=2 with M=256 in the above command?

File: experimental/delegate/mlp.mlir [hunk 5]
```diff
@@ -0,0 +1,60 @@
+// RUN: iree-compile --iree-preprocessing-transform-spec-filename=%p/mlp_spec.mlir  %s | \
+// RUN: iree-run-module --device=local-sync \
+// RUN:     --executable_plugin=$IREE_BINARY_DIR/runtime/plugins/AMD-AIE-experimental/delegate/mlp_bf16_aie_delegate$IREE_DYLIB_EXT \
+// RUN:     --module=- \
+// RUN:     --function=mlp_invocation \
+// RUN:     --input="256x256xf32=2" \
+// RUN:     --input="256x256xf32=3"
+
```
Comment (nirvedhmeshram) [inline experimental/delegate/mlp.mlir:8]: I think for there to be checks in the test you need to pipe the output to FileCHeck.

File: experimental/delegate/mlp.mlir (hunk 4 above)
Comment (daveliddell) [reply experimental/delegate/mlp.mlir:30]: The test is disabled, as Eric says ctests don't run on the hardware.  If needed, I will remove the test portion.  However, this file also serves as the source of the (manual for now) e2e test

File: experimental/delegate/mlp.mlir (hunk 5 above)
Comment (daveliddell) [reply experimental/delegate/mlp.mlir:8]: I will remove the test lines, as it's clear they are causing confusion

File: experimental/delegate/mlp.mlir (hunk 4 above)
Comment (nirvedhmeshram) [reply experimental/delegate/mlp.mlir:30]: yes makes sense to remove the CHECK-LABEL and `RUN` in front of the command.

**Comment by daveliddell** [comment]
> Fly by comment. Please use better PR descriptions (believe it or not, people are following PRs being sent to this repro, and good PR hygiene matters a lot in OSS).

@MaheshRavishankar Changed title and description, added a detailed explanation of what we're doing to the README.md.

File: experimental/delegate/mlp.mlir (hunk 4 above)
Comment (daveliddell) [reply experimental/delegate/mlp.mlir]: Done

**Review by nirvedhmeshram — APPROVED** [review]
LGTM!

**Pull request merged**